lens/lily/impl.go

```diff
@@ -612,6 +612,108 @@ func (m *LilyNodeAPI) LilySurvey(_ context.Context, cfg *LilySurveyConfig) (*sch
 	return res, nil
 }
 
+type StateReport struct {
+	Height      int64
+	TipSet      *types.TipSet
+	HasMessages bool
+	HasReceipts bool
+	HasState    bool
+}
+
+func (m *LilyNodeAPI) StateCompute(ctx context.Context, key types.TipSetKey) (interface{}, error) {
+	ts, err := m.ChainAPI.ChainGetTipSet(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	_, _, err = m.StateManager.TipSetState(ctx, ts)
+	return nil, err
+}
+
+func (m *LilyNodeAPI) FindOldestState(ctx context.Context, limit int64) ([]*StateReport, error) {
+	head, err := m.ChainHead(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var out []*StateReport
+
+	for i := int64(0); true; i++ {
+		maybeBaseTS, err := m.ChainGetTipSetByHeight(ctx, head.Height()-abi.ChainEpoch(i), head.Key())
+		if err != nil {
+			return nil, err
+		}
+
+		maybeFullTS := TryLoadFullTipSet(ctx, m, maybeBaseTS)
+		out = append(out, &StateReport{
+			Height:      int64(maybeBaseTS.Height()),
+			TipSet:      maybeBaseTS,
+			HasMessages: maybeFullTS.HasMessages,
+			HasReceipts: maybeFullTS.HasReceipts,
+			HasState:    maybeFullTS.HasState,
+		})
+		if !maybeFullTS.HasState || int64(head.Height()-abi.ChainEpoch(i)) == limit {
```

Comment:
So this is probably wrong due to null rounds

Reply:
Would it be sufficient to allow for some N number of null rounds before breaking? A value which skips null rounds and increments M until it's more than N and _then_ break? Is there another way you would approach this?